vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt

```diff
@@ -119,212 +117,137 @@ class OidcSiopWallet(
+        request: AuthenticationRequestParametersFrom,
     ): KmmResult<AuthenticationResponseResult> = catching {
-        val response = createAuthnResponseParams(request).getOrThrow()
-        request.parameters.responseType?.let {
-            if (!it.contains(ID_TOKEN) && !it.contains(VP_TOKEN))
-                throw OAuth2Exception(Errors.INVALID_REQUEST)
-                    .also { Napier.w("createAuthnResponse: Unknown response_type $it") }
-        } ?: throw OAuth2Exception(Errors.INVALID_REQUEST)
-            .also { Napier.w("createAuthnResponse: response_type null in ${request.parameters}") }
-
-        when (request.parameters.responseMode) {
-            DIRECT_POST -> authnResponseDirectPost(request, response)
-            DIRECT_POST_JWT -> authnResponseDirectPostJwt(request, response)
-            QUERY -> authnResponseQuery(request, response)
-            FRAGMENT, null -> authnResponseFragment(request, response)
-            is OTHER -> TODO()
-        }
+        AuthenticationResponseFactory(jwsService).createAuthenticationResponse(
+            request,
+            response = createAuthnResponseParams(request).getOrThrow(),
+        )
     }
 
-    private fun authnResponseDirectPost(
-        request: AuthenticationRequestParametersFrom,
-        response: AuthenticationResponse
-    ): AuthenticationResponseResult.Post {
-        val url = request.parameters.responseUrl
-            ?: request.parameters.redirectUrl
-            ?: throw OAuth2Exception(Errors.INVALID_REQUEST)
-        return AuthenticationResponseResult.Post(url, response.params.encodeToParameters())
+    /**
+     * Creates the authentication response from the RP's [params]
+     */
+    suspend fun createAuthnResponseParams(
+        params: AuthenticationRequestParametersFrom
+    ): KmmResult<AuthenticationResponse> = catching {
+        finalizeAuthorizationResponseParameters(
+            params = params,
+            preparationState = startAuthorizationResponsePreparation(params).getOrThrow(),
+        ).getOrThrow()
     }
 
     /**
-     * Per OID4VP, the response may either be signed, or encrypted (never signed and encrypted!)
+     * Starts the authorization response building process from the RP's [params]
      */
-    private suspend fun authnResponseDirectPostJwt(
-        request: AuthenticationRequestParametersFrom,
-        response: AuthenticationResponse
-    ): AuthenticationResponseResult.Post {
-        val url = request.parameters.responseUrl
-            ?: request.parameters.redirectUrl
-            ?: throw OAuth2Exception(Errors.INVALID_REQUEST)
-        val responseSerialized = buildJarm(request, response)
-        val jarm = AuthenticationResponseParameters(response = responseSerialized)
-        return AuthenticationResponseResult.Post(url, jarm.encodeToParameters())
+    suspend fun startAuthorizationResponsePreparation(
+        input: String,
+    ): KmmResult<AuthorizationResponsePreparationState> = catching {
```

Comment:
can't we  `map` the result?

Reply:
What exactly do you mean by that?
Possibly fixed in commit 2d7a99f09eaf8a8dc2c3c5885dea815299ec1ab0

Reply:
fixed in commit e0d9d1f448615405bd526b2e676580e2b90ae884